### ISSUE TITLE
[REQ] [JAVA] [resttemplate] [webclient] Improved configuration of multiple clients in a single application

### ISSUE THREAD

**Issue by jorgerod** [OpenAPITools/openapi-generator#10621]
### Is your feature request related to a problem? Please describe.

We have an application with two java clients (for example, resttemplate). Each of these clients consumes an API that is located on different servers.

The approach we're following right now is to build ourselves the multiple ApiClient beans inside @Configuration classes:

```java
    @Bean
    public RestTemplate petRestTemplate(final RestTemplateBuilder builder) {
        //config rest template...
    }

    @Bean
    public ApiClient petApiClient(final @Qualifier("petRestTemplate") RestTemplate restTemplate) {
        final ApiClient apiClient = new ApiClient(restTemplate);
        apiClient.setBasePath(restTemplate.getBasePath());
        return apiClient;
    }


    @Bean
    public PetApi petApi(final @Qualifier("petApiClient") ApiClient apiClient) {
        return new PetApi(apiClient);
    }
```

and the other client:

```java
    @Bean
    public RestTemplate otherRestTemplate(final RestTemplateBuilder builder) {
        //config other rest template...
    }

    @Bean
    public ApiClient otherApiClient(final @Qualifier("otherRestTemplate") RestTemplate restTemplate) {
        final ApiClient apiClient = new ApiClient(restTemplate);
        apiClient.setBasePath(restTemplate.getBasePath());
        return apiClient;
    }


    @Bean
    public OtherApi otherApi(final @Qualifier("otherApiClient") ApiClient apiClient) {
        return new OtherApi(apiClient);
    }
```

This has the major drawback of needing to exclude the autogenerated `ApiClient` component from the `@SpringBootApplication` or else we'll end up with duplicate beans:

```java
@SpringBootApplication
@ComponentScan(excludeFilters = @ComponentScan.Filter(type = FilterType.ASSIGNABLE_TYPE, classes = {
    org.pet.api.invoker.ApiClient.class, org.other.api.invoker.ApiClient.class
}))
public class Application {

    public static void main(final String[] args) {
        SpringApplication.run(Application.class, args);
    }

}
```

## Describe the solution you'd like

All this extra wiring wouldn't be necessary if the autogenerated classes included a @Qualifier for the RestTemplate dependency:

```java
@Autowired
public ApiClient(@Qualifier("org.pet.api.invoker.RestTemplate") RestTemplate restTemplate) {
  this.restTemplate = restTemplate;
  init();
}
```

```java
@Autowired
public ApiClient(@Qualifier("org.other.api.invoker.RestTemplate") RestTemplate restTemplate) {
  this.restTemplate = restTemplate;
  init();
}
```

This way we would only have to provide a `RestTemplate` bean with the proper name and therefore the autogenerated component could be used as is:

```java
   @Bean(name = "org.pet.api.invoker.RestTemplate")
    public RestTemplate petRestTemplate(final RestTemplateBuilder builder) {
        //config rest template...
    }

    @Bean(name = "org.other.api.invoker.RestTemplate")
    public RestTemplate otherRestTemplate(final RestTemplateBuilder builder) {
        //config other rest template...
    }
```

The only file that would require a change is the [Mustache template used by resttemplate](https://github.com/OpenAPITools/openapi-generator/blob/357f186c4e8a6b66b68eec7e823f7c70717bbaa1/modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/ApiClient.mustache#L125)

The template would look similar to this:

```java
    @Autowired
    public ApiClient(@Qualifier("{{invokerPackage}}.RestTemplate") RestTemplate restTemplate) {
        this.restTemplate = restTemplate;
        init();
    }
```

The same change could be made in webclient( [Mustache template used by webclient](https://github.com/OpenAPITools/openapi-generator/blob/357f186c4e8a6b66b68eec7e823f7c70717bbaa1/modules/openapi-generator/src/main/resources/Java/libraries/webclient/ApiClient.mustache#L120) )

**Comment by suntsa** [comment]
Another idea: if openapi-generator did not add Spring annotations (`@Component` and `@Autowired`) to `ApiClient`, it would be easier to define those beans as you please in your application. Instead of telling Spring Boot to ignore those classes in auto discovery, you would just provide the beans yourself with a few lines of code in a `@Configuration` class.

It seems this kind of feature in the moustache templates would be relatively straight forward as there are only a few places where Spring annotations are used. Some changes are also needed for configuring this behaviour but I'm not sure how would that be done in this library. Maybe with config options for the Java generator? https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/java.md#config-options

**Comment by jorgerod** [comment]
> Another idea: if openapi-generator did not add Spring annotations (`@Component` and `@Autowired`) to `ApiClient`, it would be easier to define those beans as you please in your application. Instead of telling Spring Boot to ignore those classes in ~auto discovery~ component scan, you would just provide the beans yourself with a few lines of code in a `@Configuration` class.
> 
> It seems this kind of feature in the moustache templates would be relatively straight forward as there are only a few places where Spring annotations are used. Some changes are also needed for configuring this behaviour but I'm not sure how would that be done in this library. Maybe with config options for the Java generator? https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/java.md#config-options

Hello
I just proposed that in this [issue](https://github.com/OpenAPITools/openapi-generator/issues/14832). Checking all java clients, the only one with the @Component annotation is RestTemplate.
It's a bit weird...

**Comment by suntsa** [comment]
Thanks @jorgerod ! A new issue seems like a good idea as the proposed solution is different.

**Comment by suntsa** [comment]
I guess this issue is now obsolete and can be closed due to #14832, right? ApiClient does not any more get the Spring annotations.

Btw, thanks @jorgerod for your work on the aforementioned issue! I have already taken the fix into use.

**Comment by jorgerod** [comment]
> I guess this issue is now obsolete and can be closed due to #14832, right? ApiClient does not any more get the Spring annotations.
> 
> Btw, thanks @jorgerod for your work on the aforementioned issue! I have already taken the fix into use.

Yes, i will close this issue.
Thank you
